File: rust/src/leet_code/2812.rs
// Copyright © 2010-2024 <a href="https://github.com/jerodg/">JerodG</a>
//
// This program is free software: you can redistribute it and/or modify it under the terms of the
// Server Side Public License (SSPL) as published by MongoDB, Inc., either version 1 of the License,
// or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
// even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the SSPL
// for more details.
//
// The above copyright notice and this permission notice shall be included in all copies or
// substantial portions of the Software. You should have received a copy of the SSPL along with this
// program. If not, see <a href="https://www.mongodb.com/licensing/server-side-public-license">SSPL</a>.

// Package leet_code
//
// This file contains the implementation of the Solution struct and its associated methods.
// The main purpose of this file is to solve the problem described in LeetCode 2812.
// The solution involves calculating the maximum safeness factor of a grid.

use std::collections::{BinaryHeap, VecDeque};

impl Solution {
    // NEIGHBORS is a constant that represents the possible directions in which we can move in the grid.
    const NEIGHBORS: [(i32, i32); 4] = [(1, 0), (0, 1), (-1, 0), (0, -1)];

    /// The `maximum_safeness_factor` method calculates the maximum safeness factor of a grid.
    ///
    /// # Parameters
    ///
    /// * `grid` - A mutable reference to a 2D vector of integers representing the grid.
    ///
    /// # Returns
    ///
    /// * An integer representing the maximum safeness factor of the grid.
    ///
    /// # Errors
    ///
    /// This function will panic if the grid is empty.
    pub fn maximum_safeness_factor(mut grid: Vec<Vec<i32>>) -> i32 {
        let n = grid.len();
        // The `valid_neighbors` closure returns a list of valid neighbors for a given cell in the grid.
        let valid_neighbors = |(i, j): (usize, usize)| {
            let (n, r, c) = (n as i32, i as i32, j as i32);
            Self::NEIGHBORS
                .iter()
                .map(move |(dr, dc)| (r + dr, c + dc))
                .filter(move |&(x, y)| 0 <= x && x < n && 0 <= y && y < n)
                .map(|(x, y)| (x as usize, y as usize))
        };

        // Initialize the queue with the cells that are not obstacles.
        let mut queue: VecDeque<_> = (0..n)
            .flat_map(|i| (0..n).map(move |j| (i, j)))
            .filter_map(|(i, j)| match grid[i][j] {
                0 => {
                    grid[i][j] = -1;
                    None
                }
                _ => {
                    grid[i][j] = 0;
                    Some((i, j))
                }
            })
            .collect();
        let mut alt = queue.clone();

        // Calculate the safeness factor for each cell in the grid.
        while !queue.is_empty() {
            alt.clear();
            while let Some((r, c)) = queue.pop_front() {
                alt.extend(valid_neighbors((r, c)).filter_map(|(x, y)| {
                    if grid[x][y] != -1 {
                        return None;
                    }
                    grid[x][y] = grid[r][c] + 1;
                    Some((x, y))
                }));
            }
            std::mem::swap(&mut queue, &mut alt)
        }

        // Use a binary heap to find the cell with the maximum safeness factor.
        let mut heap = BinaryHeap::from([(grid[0][0], 0, 0)]);
        while let Some((safety, i, j)) = heap.pop() {
            if i == n - 1 && j == n - 1 {
                return safety;
            }
            heap.extend(valid_neighbors((i, j)).filter_map(|(x, y)| {
                let val = grid[x][y];
                if val == -1 {
                    return None;
                }
                grid[x][y] = -1;
                Some((safety.min(val), x, y))
            }));
        }
        -1
    }
}

File: ruby/leet_code/2812.rb
# frozen_string_literal: true

# Copyright © 2010-2024 JerodG <https://github.com/jerodg/> 2010-2024.
#
# This program is free software: you can redistribute it and/or modify it under the terms of the
# Server Side Public License (SSPL) as published by MongoDB, Inc., either version 1 of the License,
# or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
# even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the SSPL
# for more details.
#
# The above copyright notice and this permission notice shall be included in all copies or
# substantial portions of the Software. You should have received a copy of the SSPL along with this
# program. If not, see <https://www.mongodb.com/licensing/server-side-public-license>.
#
# This file contains a solution for calculating the maximum safeness factor of a grid.
# It defines a DisjointSet class and a method `maximum_safeness_factor`.

# DisjointSet class represents a disjoint-set data structure.
# It provides operations for union, find and size.
class DisjointSet
  # Initializes a new instance of the DisjointSet class.
  #
  # @param [Integer] n the number of elements in the set
  def initialize(n)
    @parent = Array.new(n) { -1 }
  end

  # Finds the representative of the set that a belongs to.
  #
  # @param [Integer] a an element of the set
  # @return [Integer] the representative of the set
  def find(a)
    return a if (@parent[a]).negative?

    @parent[a] = find(@parent[a])
  end

  # Unions the sets that a and b belong to.
  #
  # @param [Integer] a an element of the set
  # @param [Integer] b an element of the set
  # @return [Integer] the representative of the unioned set
  def union(a, b)
    x = find(a)
    y = find(b)
    return x if x == y

    if size(x) < size(y)
      t = x
      x = y
      y = t
    end
    @parent[x] += @parent[y]
    @parent[y] = x
    x
  end

  # Returns the size of the set that a belongs to.
  #
  # @param [Integer] a an element of the set
  # @return [Integer] the size of the set
  def size(a)
    -@parent[find(a)]
  end
end

# Calculates the maximum safeness factor of a grid.
# The safeness factor is defined as the minimum distance to a cell with value 0.
#
# @param [Array<Array<Integer>>] grid the grid
# @return [Integer] the maximum safeness factor
def maximum_safeness_factor(grid)
  n = grid.size
  q = []
  distance = Array.new(n * n) do
    next if (grid[_1 / n][_1 % n]).zero?

    q << _1
    0
  end
  until q.empty?
    r, c = q.shift.divmod(n)
    if c < n - 1 && distance[r * n + c + 1].nil?
      distance[r * n + c + 1] = distance[r * n + c] + 1
      q << r * n + c + 1
    end
    if c.positive? && distance[r * n + c - 1].nil?
      distance[r * n + c - 1] = distance[r * n + c] + 1
      q << r * n + c - 1
    end
    if r < n - 1 && distance[(r + 1) * n + c].nil?
      distance[(r + 1) * n + c] = distance[r * n + c] + 1
      q << (r + 1) * n + c
    end
    if r.positive? && distance[(r - 1) * n + c].nil?
      distance[(r - 1) * n + c] = distance[r * n + c] + 1
      q << (r - 1) * n + c
    end
  end

  ds = DisjointSet.new(n * n)
  (0...n * n).sort_by { -distance[_1] }.each do
    ds.union(_1, 1 + _1) if _1 % n < n - 1 && distance[_1] <= distance[_1 + 1]
    ds.union(_1, _1 - 1) if (_1 % n).positive? && distance[_1] <= distance[_1 - 1]
    ds.union(_1, _1 + n) if _1 / n < n - 1 && distance[_1] <= distance[_1 + n]
    ds.union(_1, _1 - n) if (_1 / n).positive? && distance[_1] <= distance[_1 - n]
    return distance[_1] if ds.find(0) == ds.find(n * n - 1)
  end
end
